conn.go

```diff
@@ -35,3 +40,110 @@ func (c *conn) RemotePeer() peer.ID {
 func (c *conn) RemotePublicKey() ci.PubKey {
 	return c.remotePubKey
 }
+
+const (
+	recordTypeHandshake byte = 22
+	versionTLS13             = 0x0304
+	maxCiphertextTLS13       = 16384 + 256 // maximum ciphertext length in TLS 1.3
+	maxHandshake             = 65536       // maximum handshake we support (protocol max is 16 MB)
+)
+
+var errSimultaneousConnect = errors.New("detected TCP simultaneous connect")
+
+type teeConn struct {
```

Comment:
do we really need this type? It's only used by `wrappedConn`, so the logic could live there and avoid a few indirections.

Reply:
Unfortunately, we need a separate struct here, since we need to wrap the `Read` call.
The `wrappedConn` is using some helper functions like `io.ReadFull`, which then call `Read` of the underlying `io.Reader`, so we can't move the logic there.